packages/storage/src/task.ts

```diff
@@ -455,7 +455,7 @@ export class UploadTask {
    */
   on(
     type: TaskEvent,
-    nextOrObserver = undefined,
+    nextOrObserver = null,
```

Comment:
I don't know if this is the right fix. The API allows for the observer function to be undefined (https://github.com/firebase/firebase-js-sdk/blob/master/packages/storage-types/index.d.ts#L80). I think we need to update https://github.com/firebase/firebase-js-sdk/blob/master/packages/storage/src/implementation/observer.ts#L41 to accept `undefined`.

Reply:
How about changing the interface at (https://github.com/firebase/firebase-js-sdk/blob/master/packages/storage-types/index.d.ts#L80) to
```javascript
on(
...
    nextOrObserver:
      | Observer<UploadTaskSnapshot>
      | ((a: UploadTaskSnapshot) => any) = null,
...
  ): Function
```

It seems unnecessary to accept undefined if null means the same thing. Please correct me if undefined and null mean different things.

Reply:
If I understand your suggestion correctly, that's a breaking change that affects the public API.

Reply:
I take it back - we don't need to change the interface.

My point is that the change here (nextOrObserver = null) is sufficient. User can still pass `undefined` to nextOrObserver, and in this case, we will default it to `null` internally, so https://github.com/firebase/firebase-js-sdk/blob/master/packages/storage/src/implementation/observer.ts#L41  doesn't need to accept `undefined`.

Reply:
I looked through the code and the saving grace seems to be https://github.com/firebase/firebase-js-sdk/blob/master/packages/storage/src/implementation/observer.ts#L50

With that being said, I still don't think it is a good idea to diverge between out `.d.ts` file and the source. At least for the RTDB, this has caused a few issues in the past. Ideally, we would at some point auto-generate our types and that can only work if the types match the interfaces that we use in our sources.

With that being said, the current state of the PR resolves this issue and uses the type we use in index.d.ts.